### PR TITLE
M2-516: [Mobile] Activity groups: business logic

### PR DESCRIPTION
**Pull request by anq83**

### CODE REVIEW

File: src/shared/lib/utils/dateTime.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+) => {
+  const minutesElapsedFromStartOfDayForSource =
+    timeSource.hours * MINUTES_IN_HOUR + timeSource.minutes;
+  const minutesElapsedFromStartOfDayForTarget =
+    timeTarget.hours * MINUTES_IN_HOUR + timeTarget.minutes;
+  return (
+    minutesElapsedFromStartOfDayForSource >
+    minutesElapsedFromStartOfDayForTarget
+  );
+};
```
Comment (BamMironov) [inline src/shared/lib/utils/dateTime.ts:61]: To make it even more clear I'd create a MintutesComparator object, something like this:
```
MinutesComparator {
  isSourceBigger(source: Date, target: Date);
  isTargetBigger(source: Date, target: Date);
}
```
And you could use it like this `MinutesComparator(timeToCheck, intervalFrom).isSourceBigger()`

File: src/widgets/activity-group/model/factories/ActivityGroupsBuilder.ts [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+};
+
+const createActivityGroupsBuilder = (
+  inputData: ActivityGroupsBuilderInput,
+): ActivityGroupsBuilder => {
+  const { progress, appletId, allAppletActivities: activities } = inputData;
+
+  const getNow = () => new Date();
+
+  const isEventActivityInProgress = (eventActivity: EventActivity): boolean => {
```
Comment (BamMironov) [inline src/widgets/activity-group/model/factories/ActivityGroupsBuilder.ts:63]: Please move the outside of createActivityGroupsBuilder function scope

File: src/widgets/activity-group/model/factories/ActivityGroupsBuilder.ts [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+    return !!record.startAt && !record.endAt;
+  };
+
+  const getProgressRecord = (
+    eventActivity: EventActivity,
+  ): ProgressPayload | null => {
+    const record =
+      progress[appletId][eventActivity.activity.id][eventActivity.event.id];
+    return record ?? null;
+  };
```
Comment (BamMironov) [inline src/widgets/activity-group/model/factories/ActivityGroupsBuilder.ts:77]: Please move the function outside of createActivityGroupsBuilder function scope.

File: src/widgets/activity-group/model/factories/ActivityGroupsBuilder.ts [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+
+  const getStartedDateTime = (eventActivity: EventActivity): Date => {
+    const record = getProgressRecord(eventActivity)!;
+    return record.startAt!;
+  };
```
Comment (BamMironov) [inline src/widgets/activity-group/model/factories/ActivityGroupsBuilder.ts:82]: Please move the function outside of createActivityGroupsBuilder function scope.

File: src/widgets/activity-group/model/factories/ActivityGroupsBuilder.ts [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+      )!;
+      position = 1;
+    }
+
+    item.activityFlowDetails.activityPositionInFlow = position;
+    item.name = activity.name;
+    item.description = activity.description;
+    item.type = activity.type;
+    item.image = activity.image;
+  };
```
Comment (BamMironov) [inline src/widgets/activity-group/model/factories/ActivityGroupsBuilder.ts:124]: Please move the function outside of createActivityGroupsBuilder function scope.

File: src/widgets/activity-group/model/factories/ActivityGroupsBuilder.ts [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+      isTimerSet: false,
+      timeLeftToComplete: null,
+      isInActivityFlow: false,
+    };
+
+    if (isFlow) {
+      populateActivityFlowFields(item, eventActivity);
+    }
+    return item;
+  };
```
Comment (BamMironov) [inline src/widgets/activity-group/model/factories/ActivityGroupsBuilder.ts:171]: Please move the function outside of createActivityGroupsBuilder function scope.

**Review by BamMironov — CHANGES_REQUESTED** [review]

File: src/entities/event/model/operations/SheduledDateCalculator.ts [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+      return null;
+    }
+    setTime(date);
+
+    return date;
+  }
+
+  let parseInput: EventParseInput = {};
+
+  if (availability.periodicityType === PeriodicityType.Weekly) {
```
[the remark this answers is not in the document]
Comment (anq83) [reply src/entities/event/model/operations/SheduledDateCalculator.ts:70]: No sence, weekly & weekdays rely on dayspan logic below, here's only input build

File: src/shared/lib/constants/dateTime.ts [hunk 8]
```diff
@@ -0,0 +1,9 @@
+export const MS_IN_MINUTE = 60000;
+export const SECONDS_IN_MINUTE = 60;
+export const MINUTES_IN_HOUR = 60;
+
+const date = new Date();
+date.setHours(0);
+date.setMinutes(0);
+date.setSeconds(0);
+export const MIDNIGHT_OF_ANY_DATE = date;
```
[the remark this answers is not in the document]
Comment (anq83) [reply src/shared/lib/constants/dateTime.ts:9]: It depends on when the app started and was it killed

[the remark this answers is not in the document]
Comment (BamMironov) [reply src/shared/lib/constants/dateTime.ts:9]: I see. Okay, can we just name it "MIDNIGHT_DATE" then?

**Review by BamMironov — APPROVED** [review]

**Pull request merged**